extensions/basic_auth/plugin.cc

```diff
@@ -0,0 +1,192 @@
+/* Copyright 2019 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "extensions/basic_auth/plugin.h"
+
+#include <google/protobuf/io/zero_copy_stream_impl_lite.h>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_split.h"
+#include "extensions/common/json_util.h"
+
+#ifndef NULL_PLUGIN
+
+#include "extensions/metadata_exchange/base64.h"
+
+#else
+
+#include "common/common/base64.h"
+
+namespace proxy_wasm {
+namespace null_plugin {
+namespace BasicAuth {
+namespace Plugin {
+
+PROXY_WASM_NULL_PLUGIN_REGISTRY;
+
+using Base64 = Envoy::Base64;
+
+#endif
+
+static RegisterContextFactory register_BasicAuth(
+    CONTEXT_FACTORY(PluginContext), ROOT_FACTORY(PluginRootContext));
+
+struct S {
+  std::unordered_set<std::string> requestMethods;
+  std::unordered_set<std::string> encodedCredentials;
+};
+
+using ::nlohmann::json;
+using ::Wasm::Common::JsonArrayIterate;
+using ::Wasm::Common::JsonGetField;
+std::unordered_map<std::string, S> configurationData;
+
+bool PluginRootContext::onConfigure(size_t size) {
+  // Parse configuration JSON string.
+  if (size > 0 && !configure(size)) {
+    LOG_WARN("configuration has errors, but initialzation can continue.");
+  }
+  return true;
+}
+
+bool PluginRootContext::configure(size_t configuration_size) {
+  auto configuration_data = getBufferBytes(WasmBufferType::PluginConfiguration,
+                                           0, configuration_size);
+  // Parse configuration JSON string.
+  auto result = ::Wasm::Common::JsonParse(configuration_data->view());
+  if (!result.has_value()) {
+    LOG_WARN(absl::StrCat("cannot parse plugin configuration JSON string: ",
+                          configuration_data->view()));
+    return false;
+  }
+  // j is a JsonObject holds configuration data
+  auto j = result.value();
+
+  if (!JsonArrayIterate(j, "pathPrefixes", [&](const json& prefixes) -> bool {
+        auto pathPrefix =
+            JsonGetField<std::string>(prefixes, "pathPrefix").value_or("");
+        if (pathPrefix == "") return false;
+        struct S s1;
+        if (!JsonArrayIterate(prefixes, "requestMethods",
+                              [&](const json& methods) -> bool {
+                                s1.requestMethods.insert(std::string(methods));
```

Comment:
Can you use jsonValueAs to prevent undesired data type? https://github.com/istio/proxy/blob/422ba7d52af4e9e12ae75497f0d7483a565908b1/extensions/common/json_util.cc#L77

Reply:
Okok sound good, changed.